scripts/build/generate-version.ps1

```
@@ -3,14 +3,62 @@
+# CLI commitcount -> 14 bits
+function GetMSIVersionFromCLIVersion([uint32]$Major, [uint32]$Minor, [uint32]$Patch, [uint32]$CommitCount)
+{
+    if($Major -ge 0x40)
+    {
+        throw [System.NotSupportedException] "Invalid Major version - $Major. Major version must be less than 64."
+    }
+
+    if($Minor -ge 0x40)
+    {
+        throw [System.NotSupportedException] "Invalid Minor version - $Minor. Minor version must be less than 64."
+    }
+
+    if($Patch -ge 0x40)
+    {
+        throw [System.NotSupportedException] "Invalid Patch version - $Patch. Patch version must be less than 64."
+    }
+
+    if($CommitCount -ge 0x4000)
+    {
+        throw [System.NotSupportedException] "Invalid CommitCount version - $CommitCount. CommitCount version must be less than 16384."
+    }
+
+    $Major = ($Major -shl 26)
+    $Minor = ($Minor -shl 20)
+    $Patch = ($Patch -shl 14)
+    [System.UInt32]$MSIVersionNumber = ($Major -bor $Minor -bor $Patch -bor $CommitCount)
+
+    $MSIMajor = ($MSIVersionNumber -shr 24) -band 0xFF
+    $MSIMinor = ($MSIVersionNumber -shr 16) -band 0xFF
+    $MSIBuild = $MSIVersionNumber -band 0xFFFF
+    $MSIVersion = "$MSIMajor.$MSIMinor.$MSIBuild"
+
+    return $MSIVersion
+}
+
 $env:ReleaseSuffix = "dev"
 $env:MajorVersion = 1
 $env:MinorVersion = 0
 $env:PatchVersion = 0
 
-$env:CommitCountVersion = (git rev-list --count HEAD).PadLeft(6, "0")
+$CommitCount = [int32](git rev-list --count HEAD)
+$env:CommitCountVersion = ([string]$CommitCount).PadLeft(6, "0")
 
 # Zero Padded Suffix for use with Nuget
 $env:VersionSuffix = "$env:ReleaseSuffix-$env:CommitCountVersion"
 
-$env:DOTNET_CLI_VERSION = "$env:MajorVersion.$env:MinorVersion.$env:PatchVersion.$env:CommitCountVersion"
\ No newline at end of file
+$env:DOTNET_CLI_VERSION = "$env:MajorVersion.$env:MinorVersion.$env:PatchVersion.$env:CommitCountVersion"
+$env:DOTNET_MSI_VERSION = GetMSIVersionFromCLIVersion $env:MajorVersion $env:MinorVersion $env:PatchVersion $CommitCount
```

I thought about this before making this change, but I feel the other way about this. `generate-version.ps1` is a single point where we want to generate all the versions. And more importantly it is better to fail fast (example - [here](https://github.com/Sridhar-MS/cli/commit/1308e27a7426b5d384ef561fbcf8ddeeadf594a7#diff-2a391c8e75a6be63b82cfc1a9423960fR19)) when the version numbers are not within the expected range. We do not want  to be in a state where the zip/nuget has a version which is incompatible with the MSI version. But rather fail generating any artifacts if the version numbers are incompatible.